src/crimson/osd/osd_operation_sequencer.h

```diff
@@ -1,80 +1,88 @@
-  using ops_sequence_t = std::map<OpRef, seastar::promise<>>;
-
-  template <typename Func, typename HandleT, typename Result = std::invoke_result_t<Func>>
-  seastar::futurize_t<Result> start_op(
-      HandleT& handle,
-      epoch_t same_interval_since,
-      OpRef& op,
-      const spg_t& pgid,
-      Func&& func) {
-    auto& ops = pg_ops[pgid];
-    if (!op->pos) {
-      [[maybe_unused]] auto [it, inserted] = ops.emplace(op, seastar::promise<>());
-      assert(inserted);
-      op->pos = it;
+  template <typename HandleT, typename FuncT>
+  seastar::future<> start_op(HandleT& handle,
+                             epoch_t interval,
+                             uint64_t op_id,
+                             FuncT&& do_op) {
+    if (last_interval == 0) {
+      last_interval = interval;
     }
-
-    auto curr_op_pos = *(op->pos);
-    const bool first = (curr_op_pos == ops.begin());
-    auto prev_op_pos = first ? curr_op_pos : std::prev(curr_op_pos);
-    auto prev_ops_drained = seastar::now();
-    if (epoch_t prev_interval = prev_op_pos->first->interval_start_epoch;
-        !first && same_interval_since > prev_interval) {
-      // need to wait for previous operations,
-      // release the current pipepline stage
+    auto have_green_light = seastar::make_ready_future<>();
+    assert(op_id > last_op);
+    const uint64_t prev_op = std::exchange(last_op, op_id);
+    assert(interval >= last_interval);
+    if (interval > last_interval) {
+      // leave the current pipeline stage, so that the following ops can be
+      // queued up here. we cannot let the seastar scheduler to determine the
+      // order of performing these ops, once they are unblocked after the first
+      // op of the same pg interval is scheduled.
       handle.exit();
-      auto& [prev_op, prev_op_complete] = *prev_op_pos;
-      ::crimson::get_logger(ceph_subsys_osd).debug(
-          "{}, same_interval_since: {}, previous op: {}, last_interval_start: {}",
-          *op, same_interval_since, prev_op, prev_interval);
-      prev_ops_drained = prev_op_complete.get_future();
-    } else {
-      assert(same_interval_since == prev_interval || first);
+      have_green_light = unblocked.wait([prev_op, this] {
+        // wait until the previous op is unblocked
+        return last_unblocked == prev_op;
+      });
     }
-    return prev_ops_drained.then(
-      [op, same_interval_since, func=std::forward<Func>(func)]() mutable {
-      op->interval_start_epoch = same_interval_since;
-      auto fut = seastar::futurize_invoke(func);
-      auto curr_op_pos = *(op->pos);
-      curr_op_pos->second.set_value();
-      curr_op_pos->second = seastar::promise<>();
-      return fut;
+    return have_green_light.then([interval, op_id, do_op=std::move(do_op), this] {
+      auto result = seastar::futurize_invoke(do_op);
+      // unblock the next one
+      last_unblocked = op_id;
+      last_interval = interval;
+      unblocked.signal();
```

Comment:
I think there can be cases like:

1. the 2nd op and 3rd op is waiting for the 1st op;
2. the 1st op gets started and signals "unblocked";
3. the 2nd op gets started and set last_unblocked to op_id of the 2nd op;
4. the 3rd op is awaken, but as last_unblocked is changed to that of the 2ndop, it can never start.